prime-router/src/main/kotlin/tokens/TokenAuthentication.kt

```diff
@@ -207,10 +209,18 @@ class FindSenderKeyInSettings(val scope: String) : SigningKeyResolverAdapter(),
         if (!Scope.isValidScope(scope, sender)) return fail("Invalid scope for this sender: $scope")
         sender.keys.forEach { jwkSet ->
             if (Scope.scopeListContainsScope(jwkSet.scope, scope)) {
-                jwkSet.keys.forEach { jwk ->
-                    if (jwk.kid == kid) { // todo add alg test!   && jwk.alg == alg.  Or kty???
-                        return jwk.toECPublicKey() // todo implement RSA
-                    }
+                val kty = KeyType.forAlgorithm(Algorithm.parse(alg))
+
+                // find by kid and kty
+                val key = jwkSet.keys.filter { jwk -> jwk.kid == kid }
+                    .firstOrNull { jwk -> jwk.kty == kty.value }
+
+                return if (key != null && kty == KeyType.EC) {
+                    key.toECPublicKey()
+                } else if (key != null && kty == KeyType.RSA) {
+                    key.toRSAPublicKey()
+                } else {
+                    null
```

Comment:
This doesn't look quite right.   Why not add these "if" criteria to the `.filter { }` above?
For example, the user might (weirdly, I admit) have both an RSA key and a EC key.   The way this code is written now, if they are calling resolve with an EC key, but the RSA key happens to be first in to keySet, this will return null, when it should succeed.

Reply:
fixed.